Axon Spring Boot application fails to start when multiple EventUpcasterChain spring beans are defined

<!-- Please use markdown (https://guides.github.com/features/mastering-markdown/) semantics throughout the bug description. -->

### Basic information

* Axon Framework version: 4.6.1
* JDK version:  17
* Complete executable reproducer if available (e.g. GitHub Repo): See attached project zip.

### Steps to reproduce

Unzip the attached project and run:
./gradlew bootRun


### Expected behaviour
The application is expected to start successfully.

### Actual behaviour
The application fails to startup throwing this exception:
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'eventStorageEngine' defined in class path resource [upcasterchaintest/AxonConfig.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.axonframework.eventsourcing.eventstore.EventStorageEngine]: Factory method 'eventStorageEngine' threw exception; nested exception is org.axonframework.common.AxonConfigurationException: Expected single candidate for component [EventUpcasterChain]. Found candidates: [eventUpcasterChain1, eventUpcasterChain2]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:638) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:955) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.23.jar:5.3.23]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.23.jar:5.3.23]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:147) ~[spring-boot-2.7.4.jar:2.7.4]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:734) ~[spring-boot-2.7.4.jar:2.7.4]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:408) ~[spring-boot-2.7.4.jar:2.7.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[spring-boot-2.7.4.jar:2.7.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306) ~[spring-boot-2.7.4.jar:2.7.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295) ~[spring-boot-2.7.4.jar:2.7.4]
        at upcasterchaintest.ApplicationKt.main(Application.kt:14) ~[main/:na]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.axonframework.eventsourcing.eventstore.EventStorageEngine]: Factory method 'eventStorageEngine' threw exception; nested exception is org.axonframework.common.AxonConfigurationException: Expected single candidate for component [EventUpcasterChain]. Found candidates: [eventUpcasterChain1, eventUpcasterChain2]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.3.23.jar:5.3.23]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653) ~[spring-beans-5.3.23.jar:5.3.23]
        ... 19 common frames omitted
Caused by: org.axonframework.common.AxonConfigurationException: Expected single candidate for component [EventUpcasterChain]. Found candidates: [eventUpcasterChain1, eventUpcasterChain2]
        at org.axonframework.spring.config.SpringConfigurer$ComponentLocator.findBean(SpringConfigurer.java:71) ~[axon-spring-4.6.1.jar:4.6.1]
        at org.axonframework.spring.config.SpringConfigurer.defaultComponent(SpringConfigurer.java:50) ~[axon-spring-4.6.1.jar:4.6.1]
        at org.axonframework.config.DefaultConfigurer.defaultUpcasterChain(DefaultConfigurer.java:523) ~[axon-configuration-4.6.1.jar:4.6.1]
        at org.axonframework.config.Component.get(Component.java:85) ~[axon-configuration-4.6.1.jar:4.6.1]
        at org.axonframework.config.DefaultConfigurer$ConfigurationImpl.upcasterChain(DefaultConfigurer.java:1021) ~[axon-configuration-4.6.1.jar:4.6.1]
        at upcasterchaintest.AxonConfig.eventStorageEngine(AxonConfig.kt:29) ~[main/:na]
        at upcasterchaintest.AxonConfig$$EnhancerBySpringCGLIB$$67064d7b.CGLIB$eventStorageEngine$0(<generated>) ~[main/:na]
        at upcasterchaintest.AxonConfig$$EnhancerBySpringCGLIB$$67064d7b$$FastClassBySpringCGLIB$$d63c1be1.invoke(<generated>) ~[main/:na]
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.3.23.jar:5.3.23]
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-5.3.23.jar:5.3.23]
        at upcasterchaintest.AxonConfig$$EnhancerBySpringCGLIB$$67064d7b.eventStorageEngine(<generated>) ~[main/:na]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.23.jar:5.3.23]
        ... 20 common frames omitted
```

However when removing 1 of the EventUpcasterChain beans defined in src/main/kotlin/upcasterchaintest/Upcasters.kt the application will start successfully.

[upcasterchaintest.zip](https://github.com/AxonFramework/AxonFramework/files/9798517/upcasterchaintest.zip)

--- a/config/src/main/java/org/axonframework/config/DefaultConfigurer.java
+++ b/config/src/main/java/org/axonframework/config/DefaultConfigurer.java
@@ -513,15 +513,15 @@ public class DefaultConfigurer implements Configurer {
     }
 
     /**
-     * Provides the default EventUpcasterChain implementation. Subclasses may override this method to provide their own
-     * default.
+     * Provides the default {@link EventUpcasterChain} implementation, looping through all
+     * {@link #registerEventUpcaster(Function) registered} {@link EventUpcaster EventUpcasters} to collect them for a
+     * fresh {@code EventUpcasterChain}. Subclasses may override this method to provide their own default.
      *
      * @param config The configuration based on which the component is initialized.
      * @return The default EventUpcasterChain to use.
      */
     protected EventUpcasterChain defaultUpcasterChain(Configuration config) {
-        return defaultComponent(EventUpcasterChain.class, config)
-                .orElseGet(() -> new EventUpcasterChain(upcasters.stream().map(Component::get).collect(toList())));
+        return new EventUpcasterChain(upcasters.stream().map(Component::get).collect(toList()));
     }
 
     /**

--- a/config/src/test/java/org/axonframework/config/DefaultConfigurerTest.java
+++ b/config/src/test/java/org/axonframework/config/DefaultConfigurerTest.java
@@ -61,6 +61,9 @@ import org.axonframework.queryhandling.QueryHandler;
 import org.axonframework.queryhandling.QueryUpdateEmitter;
 import org.axonframework.queryhandling.SimpleQueryUpdateEmitter;
 import org.axonframework.serialization.Serializer;
+import org.axonframework.serialization.upcasting.event.EventUpcaster;
+import org.axonframework.serialization.upcasting.event.EventUpcasterChain;
+import org.axonframework.serialization.upcasting.event.IntermediateEventRepresentation;
 import org.axonframework.tracing.NoOpSpanFactory;
 import org.axonframework.tracing.SpanFactory;
 import org.junit.jupiter.api.*;
@@ -651,6 +654,55 @@ class DefaultConfigurerTest {
         verify(configurer, never()).registerCommandHandler(any());
         verify(configurer, never()).eventProcessing();
         verify(configurer, times(1)).registerQueryHandler(any());
+    }
+
+    @Test
+    void registeringNoUpcastersAndNoUpcasterChainComponentReturnsEmptyUpcasterChain() {
+        //noinspection unchecked
+        Stream<IntermediateEventRepresentation> mockStream = mock(Stream.class);
+
+        EventUpcasterChain result = DefaultConfigurer.defaultConfiguration()
+                                                     .buildConfiguration()
+                                                     .upcasterChain();
+
+        result.upcast(mockStream);
+        // Since the upcaster chain is empty, the Stream is not interacted with.
+        verifyNoInteractions(mockStream);
+    }
+
+    @Test
+    void registeringUpcastersReturnsUpcasterChainOverRulingRegisteredUpcasterChainComponent() {
+        //noinspection unchecked
+        Stream<IntermediateEventRepresentation> mockStream = mock(Stream.class);
+
+        AtomicBoolean firstUpcasterInvocation = new AtomicBoolean(false);
+        EventUpcaster firstUpcaster = mock(EventUpcaster.class);
+        when(firstUpcaster.upcast(any())).thenAnswer(it -> {
+            firstUpcasterInvocation.set(true);
+            return it.getArgument(0);
+        });
+        AtomicBoolean secondUpcasterInvocation = new AtomicBoolean(false);
+        EventUpcaster secondUpcaster = mock(EventUpcaster.class);
+        when(secondUpcaster.upcast(any())).thenAnswer(it -> {
+            secondUpcasterInvocation.set(true);
+            return it.getArgument(0);
+        });
+        EventUpcasterChain testUpcasterChain = mock(EventUpcasterChain.class);
+
+        EventUpcasterChain result = DefaultConfigurer.defaultConfiguration()
+                                                     .registerEventUpcaster(c -> firstUpcaster)
+                                                     .registerEventUpcaster(c -> secondUpcaster)
+                                                     .registerComponent(
+                                                             EventUpcasterChain.class, c -> testUpcasterChain
+                                                     )
+                                                     .buildConfiguration()
+                                                     .upcasterChain();
+
+        result.upcast(mockStream);
+
+        assertTrue(firstUpcasterInvocation.get());
+        assertTrue(secondUpcasterInvocation.get());
+        verifyNoInteractions(testUpcasterChain);
     }
 
     @SuppressWarnings("unused")

--- a/spring-boot-autoconfigure/src/test/java/org/axonframework/springboot/autoconfig/InfraConfigurationTest.java
+++ b/spring-boot-autoconfigure/src/test/java/org/axonframework/springboot/autoconfig/InfraConfigurationTest.java
@@ -95,6 +95,7 @@ class InfraConfigurationTest {
 
             result.upcast(mockStream);
 
+            // Verify map invocation first and sorted invocation second, as the EventUpcasterChain beans are ordered in that fashion.
             InOrder upcasterOrder = inOrder(mockStream);
             //noinspection ResultOfMethodCallIgnored
             upcasterOrder.verify(mockStream).map(any());

--- a/spring-boot-autoconfigure/src/test/java/org/axonframework/springboot/autoconfig/InfraConfigurationTest.java
+++ b/spring-boot-autoconfigure/src/test/java/org/axonframework/springboot/autoconfig/InfraConfigurationTest.java
@@ -1,0 +1,136 @@
+/*
+ * Copyright (c) 2010-2022. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.springboot.autoconfig;
+
+import org.axonframework.config.Configuration;
+import org.axonframework.serialization.upcasting.event.EventUpcaster;
+import org.axonframework.serialization.upcasting.event.EventUpcasterChain;
+import org.axonframework.serialization.upcasting.event.IntermediateEventRepresentation;
+import org.axonframework.spring.config.MessageHandlerLookup;
+import org.axonframework.spring.config.SpringAggregateLookup;
+import org.axonframework.spring.config.SpringAxonConfiguration;
+import org.axonframework.spring.config.SpringConfigurer;
+import org.axonframework.spring.config.SpringSagaLookup;
+import org.junit.jupiter.api.*;
+import org.mockito.*;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.EnableMBeanExport;
+import org.springframework.jmx.support.RegistrationPolicy;
+import org.springframework.test.context.ContextConfiguration;
+
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+
+/**
+ * Test class validating the behavior of the {@link InfraConfiguration}.
+ *
+ * @author Steven van Beelen
+ */
+class InfraConfigurationTest {
+
+    private ApplicationContextRunner testApplicationContext;
+
+    @BeforeEach
+    void setUp() {
+        testApplicationContext = new ApplicationContextRunner().withUserConfiguration(DefaultContext.class)
+                                                               .withPropertyValues("axon.axonserver.enabled:false");
+    }
+
+    @Test
+    void infraConfigurationConstructsSpringConfigurerAndSpringAxonConfiguration() {
+        testApplicationContext.run(context -> {
+            assertThat(context).hasSingleBean(SpringConfigurer.class);
+            assertThat(context).getBean("springAxonConfigurer")
+                               .isInstanceOf(SpringConfigurer.class);
+            assertThat(context).hasSingleBean(SpringAxonConfiguration.class);
+            // The SpringAxonConfiguration is a factory for the DefaultConfigurer.ConfigurationImpl that is private.
+            // Hence, we perform a non-null check.
+            assertThat(context).getBean("springAxonConfiguration")
+                               .isNotNull();
+        });
+    }
+
+    @Test
+    void infraConfigurationConstructsLookUpBeans() {
+        testApplicationContext.run(context -> {
+            assertThat(context).hasSingleBean(MessageHandlerLookup.class);
+            assertThat(context).hasSingleBean(SpringAggregateLookup.class);
+            assertThat(context).hasSingleBean(SpringSagaLookup.class);
+        });
+    }
+
+    @Test
+    void multipleEventUpcasterChainBeansAreCombinedInSingleEventUpcasterChainThroughConfiguration() {
+        //noinspection unchecked
+        Stream<IntermediateEventRepresentation> mockStream = mock(Stream.class);
+
+        testApplicationContext.withUserConfiguration(UpcasterContext.class).run(context -> {
+            assertThat(context).getBean("firstUpcasterChain")
+                               .isInstanceOf(EventUpcasterChain.class);
+            assertThat(context).getBean("secondUpcasterChain")
+                               .isInstanceOf(EventUpcasterChain.class);
+
+            EventUpcasterChain result = context.getBean("springAxonConfiguration", Configuration.class)
+                                               .upcasterChain();
+            assertThat(result).isNotNull();
+
+            result.upcast(mockStream);
+
+            InOrder upcasterOrder = inOrder(mockStream);
+            //noinspection ResultOfMethodCallIgnored
+            upcasterOrder.verify(mockStream).map(any());
+            //noinspection ResultOfMethodCallIgnored
+            upcasterOrder.verify(mockStream).sorted();
+        });
+    }
+
+    @ContextConfiguration
+    @EnableAutoConfiguration
+    @EnableMBeanExport(registration = RegistrationPolicy.IGNORE_EXISTING)
+    static class DefaultContext {
+
+    }
+
+    // We're not returning the result of invoking the stream operations as a simplification for adjusting the mock.
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    static class UpcasterContext {
+
+        @Order(1)
+        @Bean
+        public EventUpcasterChain firstUpcasterChain() {
+            return new EventUpcasterChain(rep -> {
+                //noinspection RedundantStreamOptionalCall
+                rep.map(ier -> ier);
+                return rep;
+            });
+        }
+
+        @Order(2)
+        @Bean
+        public EventUpcasterChain secondUpcasterChain() {
+            return new EventUpcasterChain(rep -> {
+                rep.sorted();
+                return rep;
+            });
+        }
+    }
+}

